integration/run/run_test.go

```diff
@@ -64,11 +68,94 @@ func TestVolume(t *testing.T) {
 		return obj.Status.Phase == corev1.VolumeBound &&
 			obj.Labels[labels.AcornAppName] == app.Name &&
 			obj.Labels[labels.AcornAppNamespace] == app.Namespace &&
-			obj.Labels[labels.AcornManaged] == "true"
+			obj.Labels[labels.AcornManaged] == "true" &&
+			obj.Labels[labels.AcornVolumeName] == "external-bind"
 	})
 
 	helper.WaitForObject(t, helper.Watcher(t, c), &apiv1.AppList{}, app, func(obj *apiv1.App) bool {
-		return obj.Status.Condition(v1.AppInstanceConditionParsed).Success
+		return obj.Status.Condition(v1.AppInstanceConditionParsed).Success &&
+			obj.Status.Condition(v1.AppInstanceConditionVolumes).Success
+	})
+}
+
+func TestVolumeBadClass(t *testing.T) {
+	helper.StartController(t)
+
+	ctx := helper.GetCTX(t)
+	c, _ := helper.ClientAndNamespace(t)
+
+	image, err := c.AcornImageBuild(ctx, "./testdata/volume-bad-class/Acornfile", &client.AcornImageBuildOptions{
+		Cwd: "./testdata/volume-bad-class",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	_, err = c.AppRun(ctx, image.ID, nil)
+	if err == nil {
+		t.Fatal("expected app with bad volume class to error on run")
+	}
+}
+
+func TestVolumeClassRemoved(t *testing.T) {
+	helper.StartController(t)
+
+	ctx := helper.GetCTX(t)
+	kclient := helper.MustReturn(kclient.Default)
+	c, _ := helper.ClientAndNamespace(t)
+
+	storageClasses := new(storagev1.StorageClassList)
+	err := kclient.List(ctx, storageClasses)
+	if err != nil || len(storageClasses.Items) == 0 {
+		t.Skip("No storage classes, so skipping VolumeClassRemoved")
+		return
+	}
+
+	sc := storageClasses.Items[0]
+	sc.ObjectMeta = metav1.ObjectMeta{Name: "acorn-test-custom"}
+	if err = kclient.Create(ctx, &sc); err != nil {
+		t.Fatal(err)
+	}
+	defer func() {
+		if err = kclient.Delete(context.Background(), &sc); err != nil && !apierrors.IsNotFound(err) {
+			t.Fatal(err)
+		}
+	}()
+
+	image, err := c.AcornImageBuild(ctx, "./testdata/volume-custom-class/Acornfile", &client.AcornImageBuildOptions{
+		Cwd: "./testdata/volume-custom-class",
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	app, err := c.AppRun(ctx, image.ID, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	helper.Wait(t, kclient.Watch, &corev1.PersistentVolumeClaimList{}, func(obj *corev1.PersistentVolumeClaim) bool {
+		return obj.Labels[labels.AcornAppName] == app.Name &&
+			obj.Labels[labels.AcornAppNamespace] == app.Namespace &&
+			obj.Labels[labels.AcornManaged] == "true" &&
+			obj.Labels[labels.AcornVolumeName] == "my-data"
+	})
+
+	helper.WaitForObject(t, helper.Watcher(t, c), &apiv1.AppList{}, app, func(obj *apiv1.App) bool {
+		done := obj.Status.Condition(v1.AppInstanceConditionParsed).Success &&
+			obj.Status.Condition(v1.AppInstanceConditionVolumes).Success
+		if done {
+			if err = kclient.Delete(ctx, &sc); err != nil {
```

Comment:
Is this OK to do? This is the only way I could figure out to test the scenario of "what happens when an app is using a storage class that is removed."

Reply:
The only thing that I can think of being wrong here is that you could run up against the context timeout if this takes to long to spin up. That isn't an inherent problem to this code though, more just waiting in general for our codebase. 

I say this is fine.